File: prototype/rails/src/test/actions/customers/resolve_customer_from_cookie_test.rb
require "identity_test_case"

module Customers
  class ResolveCustomerFromCookieTest < IdentityTestCase
    test "it finds the customer the cookie points at" do
      customer = create_anonymous_customer

      assert_equal customer, resolve(customer.id)
    end

    test "it follows a merge so a stale cookie lands on the verified customer" do
      anonymous = create_anonymous_customer
      verified = create_verified_customer
      CustomerMerge.create!(anonymous_customer: anonymous, customer: verified)

      assert_equal verified, resolve(anonymous.id)
    end

    test "it resolves a cookie carrying the id as a string" do
      customer = create_anonymous_customer

      assert_equal customer, resolve(customer.id.to_s)
    end

    test "it resolves nothing for a customer that no longer exists" do
      assert_nil resolve(404)
    end

    test "it resolves nothing when there is no cookie" do
      assert_nil resolve(nil)
    end

    test "it resolves nothing for a cookie that is not an id" do
      assert_nil resolve("../../etc/passwd")
    end

    test "it resolves nothing for a cookie holding a meaningless id" do
      assert_nil resolve(0)
    end

    private

    def resolve(cookie_value)
      ResolveCustomerFromCookie.new.call(cookie_value)
    end
  end
end

File: prototype/node/src/app/actions/customers/resolve-customer-from-cookie.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { buildTestApp, TEST_INSTANT } from '../../test/build-test-app.ts'
import { createAnonymousCustomer } from './create-anonymous-customer.ts'
import { resolveCustomerFromCookie } from './resolve-customer-from-cookie.ts'
import { toTimestamp } from '../../db/timestamp.ts'

test('it finds the customer the cookie names', async (t) => {
  const { db, clock, close } = await buildTestApp()
  t.after(close)
  const customer = await createAnonymousCustomer({ db, clock })

  const found = await resolveCustomerFromCookie({ db }, customer.id)

  assert.equal(found?.id, customer.id)
})

test('a cookie left holding a merged id resolves forward to the customer the history moved to', async (t) => {
  const { db, clock, close } = await buildTestApp()
  t.after(close)
  const anonymous = await createAnonymousCustomer({ db, clock })
  const verified = await createAnonymousCustomer({ db, clock })
  await db
    .insertInto('customerMerges')
    .values({
      anonymousCustomerId: anonymous.id,
      customerId: verified.id,
      createdAt: toTimestamp(TEST_INSTANT),
    })
    .execute()

  const found = await resolveCustomerFromCookie({ db }, anonymous.id)

  assert.equal(found?.id, verified.id)
})

test('it resolves nothing when the cookie named nobody', async (t) => {
  const { db, close } = await buildTestApp()
  t.after(close)

  assert.equal(await resolveCustomerFromCookie({ db }, null), null)
})

test('it resolves nothing for a customer that no longer exists', async (t) => {
  const { db, close } = await buildTestApp()
  t.after(close)

  const found = await resolveCustomerFromCookie({ db }, 999999)

  assert.equal(found, null)
})

